proto/backend.proto

```diff
@@ -21,13 +21,57 @@ message PluginConfig {
 //-----------------------------------------------
 
 service Core {
+  // Information about what a backend plugin supports/expects
+  rpc GetSchema(GetSchema.Request) returns (GetSchema.Response);
+
   // HTTP Style request
-  rpc Resource(ResourceRequest) returns (ResourceResponse);
+  rpc CallResource(CallResource.Request) returns (CallResource.Response);
 
   // Well typed query interface
   rpc DataQuery(DataQueryRequest) returns (DataQueryResponse);
 }
 
+//-----------------------------------------------
+// Schema and resource configuration
+//-----------------------------------------------
+
+message Resource {
+  // enum Permission {
+  //   VIEWER = 0;
+  //   EDITOR = 1;
+  //   ADMIN = 2;
+  // }
```

Comment:
ah -- ok, now I see the reason for all the extra routing stuff ;)

I was picturing something where:
  VIEWER : GET anything
  EDITOR+: {any verb} on any resource path

Then with enterprise we can overlay more complex role/team based authentication based on path.  But yes, there are some cases where you want VIEWER to be able to send stuff (meta-analytics, saving annotations, etc)

Rather than needing to define the routes with gRPC, is it bad to pass everything?  and let the plugin sort out how to deal with it?  most of these will be very simple.  I do think an endpoint to return an OpenAPI/swagger description could be really useful though:  https://swagger.io/specification/

Reply:
Sorry for late answer. Miss notifications of these comments.

Yes, seems like we will send the full path from Grafana to backend plugin and let the backend plugin handle it due to technical reasons for now, but we could potentially change that later if we would. 

The main thing with letting Grafana knows about resources and routes would be so that Grafana can enforce basic permissions so that plugin developer don't have to think about validating this. 2nd is that this could potentially open up for extending Grafana in more interesting ways moving forward. For example, defining an annotations resource could tell Grafana backend that it supports storing and retrieving annotations. Another example meta analytics where Grafana backend can know whether it's installed and send events to it for storage. I feel the potential is huge here for not only providing resources to the frontend, but also extend the Grafana backend using this sort of generic/dynamic api.

Reply:
I totally get why it is exciting/useful -- I am nervous about defining the routes & permissions in protobuf, and not just letting the plugin support what it needs.

If the concern is that plugin developers want a simple way to build the routes+permissions -- can't we handle that in the SDK?  without passing to grafana?  (yes we will need the user/role info passed the plugin... but we need that anyway)  Most plugins will have very simple routing needs -- for the more complex needs can we help with the SDK?  Is there urgency to solve the complex case now?

As for things like annotations....  do we need the route defined in protobuf for this?  The OpenAPI spec would give us a simple, super powerful way to expose possible features without locking down the protobuf definition.

Reply:
In general -- defining different routes for different methods and permissions feels like unnecessary overhead with little gain.  We do not need to implement macaron when what most of the time is:
```
if( method == "POST") {
   // do something
   return 'OK', 200
}
return 'Bad Request', 400 // or more specific if we really care
```